### PR TITLE
Enable zero relative timestamps to be used in log files

### LINKED ISSUE #265: candump support all timestamp formats in log files, and log format stdout
As a user, I want the option to log zero based relative timestamp formats to a file.  For instance, sometimes I don't care about absolute time, and relative is easier to read so I would like to log directly it directly to a file.  

**Issue Description**
canutils is currently unable to use timestamp type options in combination with log output -L or -l.
` -t <type>   (timestamp: (a)bsolute/(d)elta/(z)ero/(A)bsolute w date)`
and
`-l          (log CAN-frames into file.`
`-L          (use log file format on stdout)`

In regards to log output format compatibility.  Absolute, and zero logs formats work fine with can player.  So there is no reason they shouldn't be supported. 
Delta log format is different as it prints time between messages, so not suited for canplayer, unless it has a delta playback mode.  This is typically only used in specific diagnostic scenarios, I don't think we need to protect users from playing it back.

**Example usage**
```
candump -L -tz vcan0
(000.000000) vcan0 0667BAF6#89C76376D62ADD77
(000.200074) vcan0 05F541AB#
(000.400156) vcan0 187E0AA5#651E4230F28F
(000.600243) vcan0 0C091058#F45333551ABF4051

./candump -L -ta vcan0
(1607120998.528012) vcan0 0E7D3A28#0EE8651E732FBF64
(1607120998.728108) vcan0 00EDF658#48AE52014F

./candump -L -td vcan0
(000.000000) vcan0 1C7C20AE#77E569140F27AC41
(000.200083) vcan0 15E0F04D#EF
(000.200085) vcan0 008702CA#1B431B0CC811287B
```

**Fix Description**

Previously `-l if(log)` and `-L if(logformt)` bypassed all timestamp formatting by using `goto out_fflush`.  I have extracted the timestamp formatting to print_timestamp so timestamp options are now available to logging workflows.

**Known Issues**
canplayer does not support playing back -tA format logs
(2020-12-04 17:25:57.869400) vcan0 02E939C4#DE18475339

### PR DESCRIPTION
**Pull request by richnetdesign**
resolve #265

### CODE REVIEW

**Comment by richnetdesign** [comment]
Note I know the Linux style guide frowns upon use of inline directive.  It was in my previous PR, and the timestamp logic is in the hot loop so I left it for review.  I can follow up and review code generated / benchmark.  My intent was to try and preserve original performance / behavior, but reuse code.

File: candump.c [hunk 1]
```diff
@@ -221,6 +222,58 @@ static int idx2dindex(int ifidx, int socket)
...
+		break;
+	}
+}
+
+static inline void print_timestamp(const char timestamp, const struct timeval *tv, struct timeval *const last_tv)
+{
+	static char buffer[TIMESTAMPSZ];
+
+	sprint_timestamp(timestamp, tv, last_tv, &buffer[0]);
+	printf("%s",buffer);
```
Comment (hartkopp) [inline candump.c:274]: Why `&buffer[0]` ?
Shouldn't it be just `buffer` on both lines?

File: candump.c [hunk 2]
```diff
@@ -270,12 +324,16 @@ int main(int argc, char **argv)
...
+			logtimestamp = optarg[0];
 			if ((timestamp != 'a') && (timestamp != 'A') &&
 			    (timestamp != 'd') && (timestamp != 'z')) {
 				fprintf(stderr, "%s: unknown timestamp mode '%c' - ignored\n",
 					basename(argv[0]), optarg[0]);
 				timestamp = 0;
 			}
+			if ((logtimestamp != 'a') && (logtimestamp != 'z')) {
+				logtimestamp = 'a';
+			}
```
Comment (hartkopp) [inline candump.c:336]: Excellent move :-)

File: candump.c [hunk 3]
```diff
@@ -724,11 +781,13 @@ int main(int argc, char **argv)
 
 			if (log) {
 				char buf[CL_CFSZ]; /* max length */
+				char ts_buf[TIMESTAMPSZ];
+
+				sprint_timestamp(logtimestamp, &tv, &last_tv, &ts_buf[0]);
```
Comment (hartkopp) [inline candump.c:786]: just `ts_buf`again

File: candump.c [hunk 4]
```diff
@@ -76,6 +76,7 @@
 #define SOF_TIMESTAMPING_SOFTWARE (1 << 4)
 #define SOF_TIMESTAMPING_RX_SOFTWARE (1 << 3)
 #define SOF_TIMESTAMPING_RAW_HARDWARE (1 << 6)
+#define TIMESTAMPSZ 44
```
Comment (hartkopp) [inline candump.c:79]: Can you add a comment where this size estimation roughly comes from?

File: candump.c [hunk 5]
```diff
@@ -69,13 +69,14 @@
 
 /* for hardware timestamps - since Linux 2.6.30 */
 #ifndef SO_TIMESTAMPING
-#define SO_TIMESTAMPING 37
+#define SO_TIMESTAMPING 50
```
Comment (ehartko) [inline candump.c:72]: ??
This is either wrong and doesn't fit to this patch.
https://elixir.bootlin.com/linux/v5.9.13/source/include/uapi/asm-generic/socket.h#L127

Comment (richnetdesign) [reply candump.c]: Sorry that is a typo.  I'm going to need some clang-tidy and rebase workflow optimizations in my near future to stop missing easy stuff.

**Comment by hartkopp** [comment]
Fixed some whitespace issues and extended the commit message.
https://github.com/linux-can/can-utils/commit/eb66451df280f95a9a12e78b151b8d867e1b78ed
Thanks Richard!

**Pull request closed without merging**